Multiple export files

DylanHill54 opened the pull request:
# Description

Changed export graph so it exports unique files for each meter of the state graph. it also now passes the correct unit for the graph if you export the graph. For raw export it also now gives unique files for each meter and includes the native unit for each meter in the export. Also altered the logic of the file size check for download to be more accurate to the changed file format when estimating size. Also changed code so that it now uses Meter.ID whenever possible 
@bryanFowles @DylanHill54

See below for changes by @huss.

Fixes #599

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

(Describe any issues that remain or work that should still be done.)

bryanFowles committed:
export data downloads multiple files

bryanFowles committed:
adding units to csv files, currently grabbing unitId

bryanFowles committed:
adding units to csv file, pushing for dylan

bryanFowles committed:
@bryanFowles and @DylanHill54 commit with export

bryanFowles committed:
fixed format

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -9,12 +9,18 @@ import graphExport, { graphRawExport, downloadRawCSV } from '../utils/exportData
 import { FormattedMessage } from 'react-intl';
 import { TimeInterval } from '../../../common/TimeInterval';
-import { metersApi } from '../utils/api'
+import { metersApi} from '../utils/api'
```
There are formatting issues in the file. Please use the automatic file formatting in VSC (command palette, the format document) to resolve. Probably should check all files.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -9,12 +9,18 @@ import graphExport, { graphRawExport, downloadRawCSV } from '../utils/exportData
 import TooltipMarkerComponent from './TooltipMarkerComponent';
+import { State } from '../types/redux/state';
+import { useSelector } from 'react-redux';
+
```
There are extra blank lines. There should only be one.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -27,59 +33,80 @@ export default function ExportComponent(props: ExportProps) {
 	 * Passes an object containing the selected meter data to a function for export.
 	 */
+	// Meters state
+	const MetersState = useSelector((state: State) => state.meters.byMeterID);
```
Since this is a variable and to be consistent with units, do not capitalize the first word in camelcase.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -27,59 +33,80 @@ export default function ExportComponent(props: ExportProps) {
-		const data = props.exportVals.datasets;
+		const data = []
+		//using a loop to push objects one at a time into data to be organized and formated for export
+		for(let i = 0; i < props.exportVals.datasets.length; i++){
```
The current coded seems a bit convoluted and I think it is probably good to fix up. It pushes props.exportVals.datasets[i] into an array, uses it and then uses slice to get it out of the array. As a result, there is only one dataset in the array at any time. Given this, I'm not sure why it should be pushed into another array rather than setting the current index in datasets to a variable and using that. It will mean removing the array index from all the referencing code. I think the code is trying to mirror the old code but that is not important to do here. Note the date/time stamp on each exported file may vary since the readings may span different time ranges but I think that is good. The old code used the first one and that does not seem a good choice.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -27,59 +33,80 @@ export default function ExportComponent(props: ExportProps) {
 	};
 
 	const exportRawReadings = async () => {
 		if (props.selectedMeters.length === 0)
```
This was not your doing but I'm now noticing. Can you please add {} around the if statement even though it is one line. OED prefers that.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -27,59 +33,80 @@ export default function ExportComponent(props: ExportProps) {
-			const lineReading = await metersApi.rawLineReadings(props.selectedMeters, props.timeInterval);
-			downloadRawCSV(lineReading, props.defaultLanguage);
-		});
+		const data: number[] = []
```
See above about using an array as I think it also applies here but I note you do use the index to get the correct props as I would expect above.

huss commented on src/client/app/containers/ExportContainer.ts:
```diff
@@ -9,6 +9,7 @@ import { ChartTypes } from '../types/redux/graph';
 import { BarReading, LineReading, ExportDataSet } from '../types/readings';
 import { State } from '../types/redux/state';
 
+
```
Please remove the extra line that was added.

huss commented on src/client/app/containers/ExportContainer.ts:
```diff
@@ -47,7 +48,8 @@ function mapStateToProps(state: State) {
 						currentChart: chart,
-						exportVals: dataPoints
+						exportVals: dataPoints,
+						unit:  state.units.units[parseInt(state.graph.selectedUnit.toString())].name
```
This applies to all the similar uses in this file. The code takes state.graph.selectedUnit, converts to a string and then parses back to an int. Given state.graph.selectedUnit is a number, I'm unclear on why this is being done. I also note that it is a constant value so really should not be attached to the group/meter data as done here. I suspect it can be removed and then the selectedUnit can be gotten from state in ExportComponent using React hooks.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -9,12 +9,18 @@ import graphExport, { graphRawExport, downloadRawCSV } from '../utils/exportData
 interface ExportProps {
 	showRawExport: boolean;
 	selectedMeters: number[];
+	units: string;
```
I don't see where this prop is used. It seems the code uses the state to get the units using the new React hooks. Given this, I suspect this can be removed.

huss commented on src/client/app/containers/ExportContainer.ts:
```diff
@@ -145,6 +150,7 @@ function mapStateToProps(state: State) {
 		showRawExport: state.graph.chartToRender === 'line' ? true : false,
 		selectedMeters: state.graph.selectedMeters,
 		selectedGroups: state.graph.selectedGroups,
+		units: state.graph.selectedUnit.toString(),
```
There is only one unit so I'm not sure this should be plural. Other comments may mean it is going away.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -9,12 +9,18 @@ import graphExport, { graphRawExport, downloadRawCSV } from '../utils/exportData
 
 interface ExportProps {
 	showRawExport: boolean;
 	selectedMeters: number[];
```
This is a comment about the code that existed before your changes that I am noticing during my review. First, selectedGroups that is returned by ExportContainer is not used here. That was a sign to me that selecedMeters may not be needed. Looking at the code, the datasets contain the meter and group data (why the code loops over that below) and each entry has the id. (I note that raw uses the selectedMeters length which was okay since raw does not export groups.) The other uses seem to allow for getting the values from the datasets. Anyway, the state has these values so why pass via props given Redux hooks. Given all of this, I think the code should be updated to fix the issues and make it better. This is outside your changes so I suspect you will want someone else to do them.

huss commented on src/client/app/utils/exportData.ts:
```diff
@@ -9,23 +9,25 @@ import * as moment from 'moment';
 import { UserRole } from '../types/items';
 import translate from './translate';
 
+
```
Please remove all the extra blank lines.

huss commented on src/client/app/utils/exportData.ts:
```diff
@@ -66,8 +68,8 @@ export default function graphExport(dataSets: ExportDataSet[], name: string) {
 /* eslint-disable @typescript-eslint/no-unused-vars */
-export function downloadRawCSV(items: RawReadings[], defaultLanguage: string) {
-	let csvOutput = 'Label,Readings,Start Timestamp,End Timestamp\n';
+export function downloadRawCSV(items: RawReadings[], defaultLanguage: string, unit: string) {
```
It is interesting that unit is passed here but in the non-raw version it gets it in a similar way to meter with items[0]. This is something that should be looked into.

huss replied:
With my standardizing how file names are created and other info this is no longer an issue.

huss commented on src/client/app/utils/exportData.ts:
```diff
@@ -77,12 +79,13 @@ export function downloadRawCSV(items: RawReadings[], defaultLanguage: string) {
 		const startTimestamp = moment(ele.startTimestamp).format('dddd LL LTS').replace(/,/g, ''); // use regex to omit pesky commas
 		const endTimestamp = moment(ele.endTimestamp).format('dddd LL LTS').replace(/,/g, ''); // use regex to omit pesky commas
-		csvOutput += `"${ele.label}",${ele.reading},${startTimestamp},${endTimestamp}\n`; // TODO: add column for units
+		csvOutput += `"${ele.reading},${startTimestamp},${endTimestamp}\n`; // TODO: add column for units
```
The TODO should be removed as the unit is done above.

huss commented on src/client/app/utils/exportData.ts:
```diff
@@ -96,7 +99,7 @@ export function downloadRawCSV(items: RawReadings[], defaultLanguage: string) {
 // Ideally we would have a component that prompts the user and handles all the logic
 export async function graphRawExport(count: number, warningFileSize: number, fileSizeLimit: number, done: () => Promise<void>): Promise<any> {
-	const fileSize = (count * 0.0849 / 1000);
+	const fileSize = (count * 0.0857 / 1000);
```
I have not checked this but it was unexpected to me that the fileSize would get bigger given that we no longer include the id in each row. It may be correct but I want to check.

huss replied:
I checked this out and your estimate was pretty good. I update a little based on more testing.

huss reviewed (changes requested):
I thank @bryanFowles and @DylanHill54 for their first PR and addressing this issue. I've left a number of comments. Given some are general (even about previous work on export), you are welcome to discuss them with me to see how best to address them. I'm also happy to answer any questions you may have about the comments.

I did some basic testing but will not do extensive testing until changes are made. I note the following:

1. When a group is exported, the CSV has the word meter associated with it (as in meter identifier). I believe this is because meters and groups are mixed in the datasets. I don't know of an easy solution without separating the data (non-trivial change) so we might want to rename the item so it is clearer. Something to discuss to decide the best text.
2. There is a double quote (") at the start of readings in raw export in the CSV.
3. The issue that a meter's unit is not one that users normally see was known and is still an issue here. We need to come up with a scheme to put in a graphing unit or something appropriate for this. This is not part of the work of those doing this PR but is noted for completeness.

huss committed:
Merge branch 'development' into pr/DylanHill54/834

huss committed:
resolve easy edit comments

huss committed:
auto-format files

huss committed:
remove selecte meters/groups from props

huss committed:
remove units from props

huss committed:
remove " in raw CSV reading column

huss committed:
remove showRawExport from props

huss committed:
remove timeInterval from props

huss committed:
remove defaultLanguage as not actually used

huss committed:
add missing param

huss committed:
remove file sizes from state

huss committed:
use identifier to label meter, unit

huss committed:
get rid of array in graph export

- remove undefined check as does not seem needed.
- Rename function to include graph so clearer and parallel raw.

huss committed:
comments, remove undefined check

huss committed:
update dates in file names

- The end date is now the end time of last reading rather than
start time.
- Code was shifting times because it used to do multiple
meters/groups at once but removed since now single files.

huss committed:
fix up raw export

- Uses state as was recently changed for non-raw export.
- Asks once if okay to download and moved logic. Now uses window alert.
- Fixed lint for trailing ,
- User output string now says file(s) since may do more than one.
- Getting raw readings now only takes a single meter value.
- Don't have DB return meter id on raw so don't have to remove.
- Remove meter identifier from route since only one meter.

huss committed:
lint & comment

huss committed:
work on raw reading export

- Directly route DB result. Code was doing moment language setting on
the server but that did not work since the locale is on the client.
- Model does not do usual row map since no meter id and made DB use
exact identifiers desired.
- SQL changed to use desired identifiers.
- Route returns values directly rather than map and putting into new
array. Note that was also leaving the original names so it was sending
extra, unused data.
- Use short JSON identifiers so route uses less data.
- Don't route meter identifier.
- Get meter identifier from state for file name.
- Remove eslint exception since seems not needed now.

huss committed:
update file size estimate & user msg

huss committed:
export now React hooks + more

- The export container is gone and all logic is now in the component.
- The logic to do y-axis labels for line and bar graphics was refactored
into a new file as functions so can use with export
and graphics.
- The previous conversions for graph data exports was removed
and now it does it much more directly.
- The unit in the CSV now is the y-axis label so gives more
information such as the rate. The file name still uses the
unit identifier since shorter and now special characters.
- Where file names are created was standardized.

huss committed:
fix unit exported for raw

Raw export gives the values for the meter. Instead of giving the meter
unit the code gives the unit with a conversion to this since this is
the normal graphic unit. It does the special cases where the
conversion is not 1,0.

huss committed:
can import CSV that export

- Removed locale aware formatting of CSV date/time so
moment can import so upload works.
- Change how deal with header on CSV import so fact that first row has
more columns does not cause an issue.

huss committed:
line with non-hour rate now have correct values

huss committed:
fix bug if some items don't have readings

huss commented:
I have made a number of changes that are documented in the commit messages. The overview of significant changes is:

- All the comments about this PR are addressed.
- The unit given for meters is now the graphing one (when possible).
- Converted to React hooks so no container.
- Works correctly when user graphs different rates.
- Can import exported readings. Thus, the format for the CVS changed.
- The logic was simplified and extra steps were removed from older code.
- The route now sends less data.
- The end date given is the end of the last item rather than the start.

spearec commented on src/client/app/translations/data.js:
```diff
@@ -74,7 +74,7 @@ const localeData = {
 		"csv.common.param.update": "Update",
 		"csv.download.size.limit": "Sorry you don't have permissions to download due to large number of points.",
-		"csv.download.size.warning.size": "File size will be about",
+		"csv.download.size.warning.size": "File(s) size will be about (usually within 10% for large exports).",
```
This sounds a little weird, maybe "Total file size" or "File size(s)" instead?

huss replied:
Thanks for the careful reading. I agree the wording was clunky. I changed it to "Total size of all files will be about (usually within 10% for large exports)." I hope this is okay with you.

spearec reviewed (commented):
PR looks good.
I was unable to get an exported file properly imported into OED. The import worked, but had no data. I don't know if this is an issue with the code or just with my lack of knowledge on how to do this (the documentation isn't great on this). Either way, something to note here.

huss committed:
update message for export file size warning

huss commented:
As for importing and exported file, the help pages are not yet up to date with the 1.0 changes. There are also some quirky behavior that will be address later. I assuming you created the meter when you imported the readings. (If not then let me know as something else is going on.) As with several places when OED automatically creates a meter, it cannot set all the needed values so an admin must do it. I hope these steps will work for you:

1. Import on the CSV readings page with header row checked. Don't bother refreshing here since it does no good.
2. Go to the meter page and refresh. As of now, OED does not add new meters to the state so this is needed.
3. Edit the new meter to set the units. It is a little faster to graph if you also set the default graphic unit.
4. Unfortunately, meter edits do not kick the server/state. It was fixed for units & conversions but I forgot about this. Thus you need to open a shell on the web container and do: updateCikAndViews.

This gave me back the same graphic as what I exported assuming you exported the raw readings. Let me know if this works for you.

spearec reviewed (approved):
Following those steps, I was able to properly re-import the meter. I believe the issue was the updateCikAndViews part.

Since this was an issue on my part and not with the code, I have no further issues.

spearec commented:
Following those steps, I was able to properly re-import the meter. I believe the issue was the updateCikAndViews part.

Since this was an issue on my part and not with the code, I have no further issues.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/DylanHill54/834

The conflicting change was in the container but that code had
been moved to the component. Thus, the change was moved
to the new component code and the container allowed to
be deleted.

huss commented:
review by @spearec is being used to merge this PR.
